cloud-checksum/src/error.rs

```diff
@@ -84,17 +134,97 @@ impl From<serde_json::Error> for Error {
     }
 }
 
-impl<T> From<SdkError<T>> for Error
+impl From<byte_stream::error::Error> for Error {
+    fn from(err: byte_stream::error::Error) -> Self {
+        Self::IOError(io::Error::other(err))
+    }
+}
+
+/// An API error that could be returned from storage.
+#[derive(Serialize, Deserialize, Debug, Clone, Eq, PartialOrd, PartialEq, Ord, Hash)]
+pub struct ApiError {
+    /// The error kind, e.g. `AccessDenied`.
+    pub(crate) code: String,
+    /// The API call.
+    pub(crate) call: String,
+    /// The error message.
+    pub(crate) message: String,
+}
+
+impl ApiError {
+    /// Create a new error.
+    pub fn new(code: String, call: String, message: String) -> Self {
+        Self {
+            code,
+            call,
+            message,
+        }
+    }
+
+    /// Check if the error is an access denied error.
+    pub fn is_access_denied(&self) -> bool {
+        self.code == "AccessDenied"
+    }
+}
+
+impl<T> From<(&SdkError<T, HttpResponse>, String)> for ApiError
 where
-    T: Send + Sync + error::Error + CreateUnhandledError + 'static,
+    T: ProvideErrorMetadata + CreateUnhandledError + error::Error + Send + Sync + 'static,
 {
-    fn from(err: SdkError<T>) -> Self {
-        Self::AwsError(err.into_service_error().to_string())
+    fn from((err, call): (&SdkError<T, HttpResponse>, String)) -> Self {
+        Self::new(
+            err.code().unwrap_or("Unknown").to_string(),
+            call,
+            err.message()
+                .map(|msg| msg.to_string())
+                .or_else(|| err.as_service_error().map(|err| err.to_string()))
+                .unwrap_or_else(|| DisplayErrorContext(&err).to_string()),
+        )
     }
 }
 
-impl From<byte_stream::error::Error> for Error {
-    fn from(err: byte_stream::error::Error) -> Self {
-        Self::AwsError(err.to_string())
+impl Display for ApiError {
+    fn fmt(&self, f: &mut Formatter<'_>) -> fmt::Result {
+        write!(f, "{} for {}: {}", self.code, self.call, self.message)
     }
 }
+
```

Comment:
Interesting macro, @mmalenic, I'm surprised that this is necessary, I'd have expected that aws-sdk-rust would help more with error mgmt?

Reply:
It does help but it's a bit all or nothing, you can either print something like "Aws Error: unhandled" which gives no info, or the full error context with `DisplayErrorContext` which gives too much unnecessary information.

The point of this macro isn't so much about improving the error message (lines 174 - 182 of this file mostly do that), it's more to `stringify` the API call that caused the error which wouldn't be possible with generics. E.g. when an error is found, this code will show that it came from a `HeadObject` call.

For some reason, AWS errors never print what API call caused the error, they only print the error code and things like request ids if using the full context.